### PR TITLE
cleanup: refactor consensus code

### PR DESCRIPTION
**Pull request by joschisan**

### CODE REVIEW

File: fedimint-server/src/atomic_broadcast/broadcast.rs [hunk 1]
```diff
@@ -44,79 +50,95 @@ impl AtomicBroadcast {
...
-                    tracing::warn!(target: LOG_CONSENSUS,"Session with index {} has been interrupted", index);
-                }
-            });
-        }
-
-        ordered_item_receiver
+    ) -> anyhow::Result<()> {
+        // if all nodes are correct the session will take 45 to 60 seconds. The
+        // more nodes go offline the longer the session will take to complete.
+        const EXPECTED_ROUNDS_PER_SESSION: usize = 45 * 4;
```
Comment (dpc) [inline fedimint-server/src/atomic_broadcast/broadcast.rs:58]: **suggestion (non-blocking):** Units unclear. Would be best to use `Duration`, but not possible / doesn't seem like a good idea at least `*_SECS` etc. would help.

File: fedimint-server/src/consensus/mod.rs [hunk 2]
```diff
@@ -40,6 +43,8 @@ pub struct FedimintConsensus {
     pub db: Database,
     /// API for accessing state
     pub client_cfg_hash: sha256::Hash,
+    /// track the latest contribution by peer
+    pub latest_contribution_by_peer: Arc<RwLock<LatestContributionByPeer>>,
```
Comment (elsirion) [inline fedimint-server/src/consensus/mod.rs:47]: Will this be exposed via the API/UI? Or what is this data for? If not: why the `RwLock`? I hope we don't actually expect parallel writes? If not, and it is only for the UI, then maybe a [tokio watch channel](https://docs.rs/tokio/latest/tokio/sync/watch/fn.channel.html) or similar wold be appropriate?

**Review by elsirion — APPROVED** [review]
Looks mostly good, commits could be a bit better organized/named, but works for me, AlephBFT is pretty much your playground for now ^^

File: fedimint-core/src/block.rs [hunk 3]
```diff
@@ -1,42 +1,40 @@
...
 
 impl Block {
     /// A blocks header consists of 40 bytes formed by its index in big endian
     /// bytes concatenated with the merkle root build from the consensus
-    /// hashes of its [OrderedItem]s or 32 zero bytes if the block is
+    /// hashes of its [AcceptedItem]s or 32 zero bytes if the block is
     /// empty. The use of a merkle tree allows for efficient inclusion
     /// proofs of accepted consensus items for clients.
-    pub fn header(&self) -> [u8; 40] {
+    pub fn header(&self, index: u64) -> [u8; 40] {
```
[the remark this answers is not in the document]
Comment (joschisan) [reply fedimint-core/src/block.rs:34]: Agree. Deferred to follow up.

File: fedimint-core/src/block.rs [hunk 4]
```diff
@@ -1,42 +1,40 @@
...
 
 /// All items ordered in a session that have been accepted by Fedimint Consensus
 /// are recorded in the corresponding block. A running Federation produces a
 /// [Block] roughly every five minutes.  Therefore, just like in Bitcoin, a
 /// [Block] might be empty if no items are ordered in that time or all ordered
 /// items are discarded by Fedimint Consensus.
-#[derive(Clone, Debug, PartialEq, Eq, Encodable, Decodable, Deserialize, Serialize)]
+#[derive(Clone, Debug, PartialEq, Eq, Encodable, Decodable)]
 pub struct Block {
-    pub index: u64,
```
[the remark this answers is not in the document]
Comment (joschisan) [reply fedimint-core/src/block.rs:26]: One calls the header to verify the signatures. If the block had its own index to calculate the header one could forget to check wether the block has the correct index.

File: fedimint-server/Cargo.toml [hunk 5]
```diff
@@ -45,7 +45,8 @@ tokio-stream = "0.1.11"
 tokio-rustls = "0.23.4"
 tokio-util = { version = "0.7.4", features = [ "codec" ] }
 tracing-subscriber = { version = "0.3.16", features = [ "env-filter" ] }
-aleph-bft = { version = "0.20.5", default-features = false }
+aleph-bft = { version = "0.30.0", default-features = false }
+aleph-bft-types = "0.10.0"
```
[the remark this answers is not in the document]
Comment (joschisan) [reply fedimint-server/Cargo.toml:49]: Unfortunately not.

File: fedimint-server/src/consensus/mod.rs (hunk 2 above)
Comment (joschisan) [reply fedimint-server/src/consensus/mod.rs:47]: This is exposed via the api to flag peers as lagging. Did not choose the RwLock for this, we do not have parallel writes though. I was thinking to use a database entry to track this to ensure consistency across shutdowns

**Pull request merged**